File: String/reverseStr.py
"""
Write a function that takes a string as input and returns the string reversed.

Example:
Given s = "hello", return "olleh".
"""

class Solution(object):
    def reverseString_build(self, s):
        """
        :type s: str
        :rtype: str
        """
        return s[::-1]

    def reverseString(self, s):
        """
        :type s: str
        :rtype: str
        """
        n = len(s)
        if n <= 1:
            return s
        else:
            s_lt = list(s)
            for i in xrange(n//2):
                s_lt[i], s_lt[-i-1] = s_lt[-i-1], s_lt[i]
            return "".join(s_lt)



if __name__ == '__main__':
    sol = Solution()
    print sol.reverseString_build('hello')
    print sol.reverseString('world')

File: String/reverseStr.cpp
/*
 Write a function that takes a string as input and returns the string reversed.
 
 Example:
 Given s = "hello", return "olleh".
*/

#include <iostream>
#include <string>

using std::string;
using std::cout;
using std::endl;

class Solution {
public:
    string reverseString(string s) {
        size_t n = s.size();
        if (n <= 1)
        {
            return s;
        }
        else
        {
            int n_f = int(n*0.5);
            for (int i = 0; i < n_f; ++i)
            {
                char temp = s.at(i);
                s.at(i) = s.at(n-1-i);
                s.at(n-1-i) = temp;
            }
            return s;
        }
    }
};


int main()
{
    string inp_str = string("Hello World!");
    Solution sol = Solution();
    cout << sol.reverseString(inp_str) << endl;

    return 0;
}
